sgkit_vcf/utils.py

```diff
@@ -0,0 +1,55 @@
+import itertools
+import struct
+from pathlib import Path
+from typing import IO, Any, Iterator, Optional, Sequence, TypeVar
+
+T = TypeVar("T")
+
+
+def ceildiv(a: int, b: int) -> int:
+    """Safe integer ceil function"""
+    return -(-a // b)
+
+
+# https://dev.to/orenovadia/solution-chunked-iterator-python-riddle-3ple
+def chunks(iterator: Iterator[T], n: int) -> Iterator[Iterator[T]]:
+    """
+    Convert an iterator into an iterator of iterators, where the inner iterators
+    each return `n` items, except the last, which may return fewer.
+    """
+
+    for first in iterator:  # take one item out (exits loop if `iterator` is empty)
+        rest_of_chunk = itertools.islice(iterator, 0, n - 1)
+        yield itertools.chain([first], rest_of_chunk)  # concatenate the first item back
+
+
+def get_file_length(path: Path) -> int:
+    """Get the length of a file in bytes."""
+    return path.stat().st_size
+
+
+def get_file_offset(vfp: int) -> int:
+    """Convert a block compressed virtual file pointer to a file offset."""
+    address_mask = 0xFFFFFFFFFFFF
+    return vfp >> 16 & address_mask
+
+
+def read_bytes(
+    f: IO[Any], fmt: str, nodata: Optional[Sequence[Any]] = None
+) -> Sequence[Any]:
+    """Read bytes using a `struct` format string"""
+    data = f.read(struct.calcsize(fmt))
+    if not data:
+        return nodata or ()
+    return struct.Struct(fmt).unpack(data)
+
+
+def at_eof(f: IO[Any]) -> bool:
```

Comment:
I'm slightly worried about this function - doesn't this mean that we can't read from a stream, etc? Wouldn't it be simpler to ``assert(len(f.read()) == 0)``  in the calling function? Any time I depend on ``seek`` and ``tell`` I end up regretting it!

Reply:
Fixed - much more idiomatic. Thanks!

https://github.com/pystatgen/sgkit-vcf/pull/1/commits/ced553f861e816b9c0370f5e61b4a4662b475822